### PR TITLE
clear SSA field managers from composed resources when migrating to composition functions

### PR DESCRIPTION
**Pull request by ravilr**
### Description of your changes

- Fixes https://github.com/crossplane/crossplane/issues/5639

The newly added e2e test fails without the changes around managedFields Upgrade() in this PR.  With the Upgrade() change, it passes.

@negz PTAL.

I have: 

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~~Added or updated unit tests.~~
- [X] Added or updated e2e tests.
- [ ] ~~Linked a PR or a [docs tracking issue] to [document this change].~~
- [ ] ~~Added `backport release-x.y` labels to auto-backport this PR.~~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Comment by ravilr** [comment]
@crossplane/crossplane-maintainers looks like `TestCompositionRealtimeRevisionSelection` has been flaky for a while, doesn't look related to this change.  the wait time for `ClaimHasPatchedField` assertion there could use a bit longer timeout than the current 10 sec.

File: internal/controller/apiextensions/composite/composition_functions.go [hunk 1]
```diff
@@ -825,3 +858,86 @@ func UpdateResourceRefs(xr resource.ComposedResourcesReferencer, desired Compose
...
+//
+// Step 1: All fields are owned by manager 'crossplane' operation 'Update'. This
+// represents all fields set by the XR controller up to this point.
+//
+// Step 2: Upgrade is called for the first time. We clear all field managers.
+//
+// Step 3: The XR controller server-side applies its fully specified intent
+// as field manager with prefix 'apiextensions.crossplane.io/composed/'. This becomes the
+// manager of all the fields that are part of the XR controller's fully
+// specified intent. All existing fields the XR ontroller didn't specify
```
Comment (jbw976) [inline internal/controller/apiextensions/composite/composition_functions.go:889]: nit:
```suggestion
// specified intent. All existing fields the XR controller didn't specify
```

File: internal/controller/apiextensions/composite/composition_functions.go [hunk 2]
```diff
@@ -825,3 +858,86 @@ func UpdateResourceRefs(xr resource.ComposedResourcesReferencer, desired Compose
...
+	foundBFA := false
+	idxBFA := -1
+
+	for i, e := range obj.GetManagedFields() {
+		if strings.HasPrefix(e.Manager, FieldOwnerComposedPrefix) {
+			foundSSA = true
+		}
+		if e.Manager == "before-first-apply" {
+			foundBFA = true
+			idxBFA = i
```
Comment (jbw976) [inline internal/controller/apiextensions/composite/composition_functions.go:911]: is it possible to have multiple managers of type `before-first-apply`? if it is possible (i genuinely don't know), do we need to keep track of each of their index?

File: internal/controller/apiextensions/composite/composition_functions.go [hunk 3]
```diff
@@ -825,3 +858,86 @@ func UpdateResourceRefs(xr resource.ComposedResourcesReferencer, desired Compose
...
+		]`, idxBFA, obj.GetResourceVersion()))
+		return errors.Wrap(resource.IgnoreNotFound(u.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, p))), "cannot remove before-first-apply from field managers")
+
+	// We didn't find our SSA field manager or the before-first-apply field
+	// manager. This means we haven't started the upgrade. The first thing we
+	// want to do is clear all managed fields. After we do this we'll let our
+	// SSA field manager apply the fields it cares about. The result will be
+	// that our SSA field manager shares ownership with a new manager named
+	// 'before-first-apply'.
+	default:
```
Comment (jbw976) [inline internal/controller/apiextensions/composite/composition_functions.go:936]: is it possible that we could reach this `default` case without finding our SSA field manager but yes we did find the before-first-apply manager? Is that valid, and if so, this operation is OK for that case too?

File: test/e2e/apiextensions_test.go [hunk 4]
```diff
@@ -377,6 +378,51 @@ func TestPropagateFieldsRemovalToXRAfterUpgrade(t *testing.T) {
...
+				funcs.ApplyClaim(FieldManager, manifests, "claim.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "claim.yaml"),
+				funcs.ResourcesHaveConditionWithin(5*time.Minute, manifests, "claim.yaml", xpv1.Available()),
+			)).
+			Assess("ConvertToPipelineCompositionUpgrade", funcs.ApplyResources(FieldManager, manifests, "composition-xfn.yaml")).
+			Assess("UpdateClaim", funcs.ApplyClaim(FieldManager, manifests, "claim-update.yaml")).
+			Assess("FieldsRemovalPropagatedToMR",
+				// field removals are propagated claim -> XR -> MR, after converting composition from native to pipeline mode
+				funcs.ComposedResourcesHaveFieldValueWithin(1*time.Minute, manifests, "claim.yaml",
+					"spec.forProvider.fields.tags[newtag]", funcs.NotFound,
```
Comment (jbw976) [inline test/e2e/apiextensions_test.go:411]: is it also worth a check that `tags[tag]` still exists, to confirm it wasn't accidentally removed? Up to you if you think that's a useful check to add :)

**Review by jbw976 — COMMENTED** [review]
Awesome @ravilr thank you so much for taking this on for the v1.16 milestone still! Really appreciate the effort 💪 

this looks pretty reasonable to me, but I'm not an expert on the subject matter - I've added a few implementation questions that I doubt are issues, but may be good just to confirm in our thinking. Thanks! 🙇‍♂️

File: internal/controller/apiextensions/composite/composition_functions.go (hunk 1 above)
Comment (ravilr) [reply internal/controller/apiextensions/composite/composition_functions.go]: fixed, thanks.

File: internal/controller/apiextensions/composite/composition_functions.go (hunk 2 above)
Comment (ravilr) [reply internal/controller/apiextensions/composite/composition_functions.go:911]: no, i don't think a K8s resource `metadata.managedFields`  could end up with multiple managedField entries with manager as `before-first-apply`.

File: internal/controller/apiextensions/composite/composition_functions.go (hunk 3 above)
Comment (ravilr) [reply internal/controller/apiextensions/composite/composition_functions.go:935]: >  if so, this operation is OK for that case too?

yes, it is OK/safe to clear all managedField entries, even if `foundBFA && !foundSSA`

File: test/e2e/apiextensions_test.go (hunk 4 above)
Comment (ravilr) [reply test/e2e/apiextensions_test.go:411]: Updated to assert that both field removals and updates (tags[tag)) propagates as expected.

**Review by negz — APPROVED** [review]
Looks great, thanks for tackling this @ravilr!

**Pull request merged**